source-alpaca/pull.go

```diff
@@ -0,0 +1,156 @@
+		}
+	}
+
+	var resources []resource
+	for _, binding := range open.Open.Capture.Bindings {
+		var res resource
+		if err := pf.UnmarshalStrict(binding.ResourceSpecJson, &res); err != nil {
+			return fmt.Errorf("parsing resource config: %w", err)
+		}
+		res = resourceWithConfigDefaults(res, cfg)
+
+		res.startDate = cfg.StartDate
+		// If we have persisted a checkpoint indicating progress for this resource, use that instead
+		// of the configured startDate.
+		if got, ok := checkpoint.BackfilledUntil[res.Name]; ok {
+			res.startDate = got
+			log.WithFields(log.Fields{
+				"Name":      res.Name,
+				"StartDate": res.startDate,
+			}).Info("set resource StartDate from checkpoint")
+		}
+
+		resources = append(resources, res)
+	}
+
+	var capture = &capture{
+		Stream:   stream,
+		Bindings: resources,
+		Config:   cfg,
+		State:    checkpoint,
+		Output:   &boilerplate.PullOutput{Stream: stream},
+	}
+	return capture.Run()
+}
+
+type capture struct {
+	Stream   pc.Driver_PullServer
+	Bindings []resource
+	Config   config
+	State    captureState
+	Output   *boilerplate.PullOutput
+}
+
+func (c *capture) Run() error {
+	// Notify Flow that we're starting.
+	if err := c.Output.Ready(); err != nil {
+		return err
+	}
+
+	var eg, ctx = errgroup.WithContext(c.Stream.Context())
+
+	// Capture each binding.
+	for idx, r := range c.Bindings {
+		idx, r := idx, r // Copy the loop variables for each closure
+		eg.Go(func() error {
+			return c.CaptureBinding(ctx, idx, r)
+		})
+	}
+
+	return eg.Wait()
+}
+
+func (c *capture) CaptureBinding(ctx context.Context, bindingIdx int, r resource) error {
+	dataClient := marketdata.NewClient(marketdata.ClientOpts{
+		ApiKey:    c.Config.ApiKeyID,
+		ApiSecret: c.Config.ApiSecretKey,
+	})
+
+	streamClient := marketdataStream.NewStocksClient(
+		r.Feed,
+		marketdataStream.WithCredentials(c.Config.ApiKeyID, c.Config.ApiSecretKey),
+	)
+
+	worker := alpacaWorker{
+		output:       c.Output,
+		bindingIdx:   bindingIdx,
+		dataClient:   dataClient,
+		streamClient: streamClient,
+		resourceName: r.Name,
+		symbols:      r.GetSymbols(),
+		feed:         r.Feed,
+		freePlan:     c.Config.Advanced.IsFreePlan,
+	}
+
+	eg, ctx := errgroup.WithContext(ctx)
+
+	caughtUp := make(chan struct{})
+
+	if !c.Config.Advanced.DisableBackfill {
+		eg.Go(func() error {
+			return worker.backfillTrades(ctx, r.startDate, c.Config.Advanced.StopDate, c.Config.effectiveMaxBackfillInterval, c.Config.effectiveMinBackfillInterval, caughtUp)
+		})
+	} else {
+		// If backfilling is disabled, there's nothing to catch up on. Streaming can start right away.
+		close(caughtUp)
+	}
+
+	if !c.Config.Advanced.DisableRealTime {
+		eg.Go(func() error {
+			// Wait until the backfilling is caught up (or disabled) before starting streaming.
+			// Throughput will most likely be limited by the network or journal append limits. It
+			// may be possible that there is such a large amount of data that the backfilling will
```

Comment:
Is it possible to log a warning/error in the event that backfilling looks like it will probably never catch up? I haven't worked through all the edge cases but something like:

```go
if itemDelay := time.Since(item.Trade.Timestamp); itemDelay > (prevDelay + time.Hour) {
  log.WithField("delay", itemDelay.String()).Warn("backfill delay increasing")
  prevDelay = itemDelay
} else if itemDelay < prevDelay {
  prevDelay = itemDelay
}
```

That is, keep track of some rough estimate of "how far in the past we are" just by looking at each trade document, and log a warning if that number appears to be increasing substantially over time.

Reply:
Seems like a good idea - I'll add logging so that if a backfill cycle takes longer to complete than the actual data period it covers, there will be a warning message emitted. This would be indicative of the backfill being unable to catch up.